### PR TITLE
163383136 Add functionality for user profile retrieval and update

### PR DESCRIPTION
**Pull request by oma0256**
**Description**
One of the most critical features of this application is profiles. This allows a user to view personal information about other users such as what articles they have published, who they follow, who follows them.

**Type of change**
Feature

**How has this been tested**
Unit tests

**Relevant Pivotal Tracker Stories**
[#163383136](https://www.pivotaltracker.com/n/projects/2240000/stories/163383136)

### CODE REVIEW

**Review by missvicki — CHANGES_REQUESTED** [review]
The UI still needs some tricking to do. I also noticed that when I edit my profile. The firstname and last name is not returned to me or a success message to tell me I edited the profile. Other than that, work looks good.

**Comment by karuhanga** [comment]
Good work on this @oma0256.
Would it be possible to add more space between the username and this button, or better yet move it to the right? It seems kind of squashed.
image

File: src/__mocks__/axios.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
 export default { 
     post : jest.fn(() => Promise.resolve({data: {}})),
-    get : jest.fn(() => Promise.resolve({data: {}})) }
\ No newline at end of file
+    get : jest.fn(() => Promise.resolve({data: {}})),
+    patch : jest.fn(() => Promise.resolve({data: {}})) }
```
Comment (collinewait) [inline src/__mocks__/axios.js]: EOF

File: src/assets/styles/Profile.css [hunk 2]
```diff
@@ -0,0 +1,22 @@
+.FileUploaderContainer {
```
Comment (collinewait) [inline src/assets/styles/Profile.css]: File extension should be `.scss`

File: src/containers/Profiles/EditProfileContainer.test.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    });
+
+    it("mapDispatchToProps", () => {
+        const dispatch = jest.fn();
+        mapDispatchToProps(dispatch).fetchUserProfile();
+        mapDispatchToProps(dispatch).requestEditProfile();
+        mapDispatchToProps(dispatch).dismissMessage();
+        expect(dispatch.mock.calls.length).toBe(3);
+    });
+})
```
Comment (collinewait) [inline src/containers/Profiles/EditProfileContainer.test.js]: EOF

File: src/components/Navbar.jsx [hunk 4]
```diff
@@ -1,13 +1,39 @@
...
+    props.fetchUserProfile({ username, token });
+};
+
+const methods = {
+    componentDidMount
+};
 
 export const Navbar = (props) => {
-    const { navBar } = props;
+    const { profile } = props;
```
Comment (collinewait) [inline src/components/Navbar.jsx]: We can use destrustructuring  directly where `props` is in the function.
`const Navbar = ({profile}) => {...}`. After, we can get rid of line `23`

File: src/components/Navbar.jsx [hunk 5]
```diff
@@ -62,8 +86,20 @@ export const Navbar = (props) => {
...
 
+const mapDispatchToProps = dispatch => {
+    return {
+        fetchUserProfile: payload => dispatch(fetchUserProfile(payload))
+    }
+}
+
+Navbar.propTypes = {
+    profile: PropTypes.object.isRequired,
+    fetchUserProfile: PropTypes.func.isRequired
```
Comment (collinewait) [inline src/components/Navbar.jsx]: `fetchUserProfile` is not a prop for NavBar. It is safe to remove it from prop validation.

File: src/components/Profiles/EditProfile.jsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        firstName, 
+        lastName, 
+        bio, 
+        profileLoading, 
+        loadingEditButton, 
+        disabled 
+    } = props;
+
+    return (
+        <Container >
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: We shouldn't have space in `<Container>`

File: src/components/Profiles/EditProfile.jsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                        <Form onSubmit={onSubmitHandler}>
+                            <Container textAlign="left">
+                                <div className='spaceComponents'>
+                                    <Form.Input 
+                                        fluid 
+                                        placeholder='First Name' 
+                                        value={firstName} 
+                                        name="firstName" 
+                                        onChange={onChangeHandler}
+                                        />
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/EditProfile.jsx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                                    <Form.Input 
+                                        fluid 
+                                        placeholder='Last Name' 
+                                        value={lastName} 
+                                        name="lastName" 
+                                        onChange={onChangeHandler}
+                                        />
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/EditProfile.jsx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                                    <textarea 
+                                        placeholder='Bio' 
+                                        value={bio} 
+                                        name="bio" 
+                                        onChange={onChangeHandler}
+                                        />
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/EditProfile.jsx [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                                    <Button
+                                        basic
+                                        color='teal'
+                                        type='submit'
+                                        loading={loadingEditButton}
+                                        disabled={disabled}
+                                        className="UpdateButton"
+                                        >
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: The closing end `>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/EditProfile.jsx [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                        </Form>
+                    </GridColumn>
+                    <GridColumn>
+                        <Container textAlign="right">
+                            <Image 
+                                src={profile.image} 
+                                size='small' 
+                                avatar 
+                                style={{ borderRadius: "50%" }} 
+                                />
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/EditProfile.jsx [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                            <div className="FileUploaderContainer">
+                                <input 
+                                    className="FileUploader" 
+                                    type='file' 
+                                    name="image" 
+                                    onChange={imageHandler}
+                                    /><br/><br/>
```
Comment (collinewait) [inline src/components/Profiles/EditProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.
Place each `<br />` on a new line and add a space between `br` and `/` in `<br />`

File: src/components/Profiles/FollowerFollowing.jsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import { List, Image, Container } from 'semantic-ui-react';
+import { Link } from 'react-router-dom';
+import PropTypes from 'prop-types';
+
+const FollowerFollowing = props => {
+    const { location, profile, match } = props;
```
Comment (collinewait) [inline src/components/Profiles/FollowerFollowing.jsx]: `location` and `match`  missing in prop validation.

File: src/components/Profiles/FollowerFollowing.jsx [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    if (pathname === `/${username}/following`){ 
+        title = `${username} Following`;
+        noUsers = `${username} is not following anybody yet`;
+    }
+
+    return (
+        <Container centered>
+            <h1>{title}</h1>
+            {!users || users.length === 0 ? <h3>{noUsers}</h3> : users.map(user => (
+                <Container>
```
Comment (collinewait) [inline src/components/Profiles/FollowerFollowing.jsx]: Add a `key` prop that uniquely identifies each element.

File: src/components/Profiles/UserProfile.jsx [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+                                <span>{profile.number_of_following} Following</span>
+                            </Link>
+                        </Grid.Column>
+                        <Grid.Column width={4} floated="right">
+                            <Image 
+                                src={profile.image} 
+                                size='small' 
+                                avatar 
+                                style={{ borderRadius: "50%" }}
+                                />
```
Comment (collinewait) [inline src/components/Profiles/UserProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/UserProfile.jsx [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+import { connect } from 'react-redux'
+import '../../assets/styles/Profile.css';
+
+export const UserProfile = props => {
+    const { profileState, articlesState, onDismissHandler } = props;
+    const { profile, profileUpdateMessage } = profileState;
+    const { articles } = articlesState;
+    const username = localStorage.getItem("username");
+    const editButton = username === profile.username ? 
+        <Link to={`/${profile.username}/edit`}>
```
Comment (collinewait) [inline src/components/Profiles/UserProfile.jsx]: Add parentheses`()` around jsx. i.e
Screenshot 2019-03-22 at 22 59 54

 Do the same on the `Message` tag with proper indentation as well in both cases..

File: src/components/Profiles/UserProfile.jsx [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+                                avatar 
+                                style={{ borderRadius: "50%" }}
+                                />
+                        </Grid.Column>
+                    </Grid.Row>
+                    <ArticleList 
+                        articles={articles} 
+                        title="My Articles" 
+                        paginating
+                        />
```
Comment (collinewait) [inline src/components/Profiles/UserProfile.jsx]: The closing end `/>` should be on the same line of the opening tag `<`.

File: src/components/Profiles/UserProfile.jsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        <Segment loading={profileState.loading || articlesState.loading}>
+            <Container text>
+                {successfulProfileUpdate}
+                <Grid>
+                    <Grid.Row>
+                        <Grid.Column width={9} className="ProfileEdit">
+                            <h2>{name}</h2>
+                            {editButton}
+                            <h3>{profile.bio}</h3>
+                            <Link to={`/${profile.username}/followers`}>
```
Comment (collinewait) [inline src/components/Profiles/UserProfile.jsx:48]: Improve `span` tags as shown below.
Screenshot 2019-03-22 at 22 57 42

File: src/components/Profiles/UserProfile.jsx [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    Grid, 
+    Image, 
+    Button, 
+    Segment, 
+    Message 
+} from 'semantic-ui-react';
+import { Link } from 'react-router-dom';
+import ArticleList from '../article/ArticleList';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux'
```
Comment (collinewait) [inline src/components/Profiles/UserProfile.jsx]: Terminate this statement(`;`)

Comment (collinewait) [inline src/components/Profiles/UserProfile.jsx]: Arrange imports. Third party libraries should be imported first.

File: src/containers/Profiles/EditProfileContainer.jsx [hunk 20]
```diff
@@ -0,0 +1,98 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { 
+    fetchUserProfile, 
+    requestEditProfile 
+} from '../../store/actions/async/profiles';
+import EditProfile from '../../components/Profiles/EditProfile';
+import { dismissMessage } from '../../store/actions/sync/profiles';
+import PropTypes from 'prop-types';
```
Comment (collinewait) [inline src/containers/Profiles/EditProfileContainer.jsx]: Arrange imports. Third party libraries should be imported first.

File: src/containers/Profiles/EditProfileContainer.jsx [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+            onChangeHandler: this.onChangeHandler,
+            imageHandler: this.imageHandler,
+            onSubmitHandler: this.onSubmitHandler,
+            onDismissHandler: this.onDismissHandler,
+            ...this.props
+        };
+
+        return <EditProfile {...editProfileProps}/>
+    };
+};
```
Comment (collinewait) [inline src/containers/Profiles/EditProfileContainer.jsx]: unnecessary termination.

File: src/containers/Profiles/EditProfileContainer.jsx [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+    state = {
+        firstName: "",
+        lastName: "",
+        bio: "",
+        image: null,
+        disabled: true
+    };
+
+    componentDidMount = () => {
+        const { username } = this.props.match.params;
```
Comment (collinewait) [inline src/containers/Profiles/EditProfileContainer.jsx]: `username` and `match` missing in prop validation. 
We can also use destructuring like:
` const { match: { params: {username} } = { params: {} } } = this.props;`

Use destructuring to get items on `this.props`. `this.props` is duplicated in this file. Make the appropriate changes.

File: src/containers/Profiles/EditProfileContainer.jsx [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+        const editProfileProps = {
+            ...this.state,
+            onChangeHandler: this.onChangeHandler,
+            imageHandler: this.imageHandler,
+            onSubmitHandler: this.onSubmitHandler,
+            onDismissHandler: this.onDismissHandler,
+            ...this.props
+        };
+
+        return <EditProfile {...editProfileProps}/>
```
Comment (collinewait) [inline src/containers/Profiles/EditProfileContainer.jsx]: Terminate(`;`) this line and add a space before `/>`

File: src/containers/Profiles/FollowersFollowing.jsx [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+export class FollowersFollowing extends Component {
+    componentDidMount = () => {
+        const { username } = this.props.match.params;
+        const token = localStorage.getItem("token");
+        this.props.fetchUserProfile({ username, token });
+    };
+
+    render() {
+        return <FollowerFollowing {...this.props} />
+    };
```
Comment (collinewait) [inline src/containers/Profiles/FollowersFollowing.jsx]: unnecessary termination on render.

File: src/containers/Profiles/EditProfileContainer.jsx [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+    };
```
Comment (collinewait) [inline src/containers/Profiles/EditProfileContainer.jsx]: unnecessary termination on render.

File: src/containers/Profiles/UserProfileContainer.jsx [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+import { connect } from 'react-redux';
+import { fetchUserProfile } from '../../store/actions/async/profiles';
+import { fetchArticlesAction } from '../../store/actions/async/ArticleActions';
+import UserProfile from '../../components/Profiles/UserProfile';
+import { dismissMessage } from '../../store/actions/sync/profiles';
+import PropTypes from 'prop-types';
+
+export class UserProfileContainer extends Component {
+    fetchProfile = () => {
+        const { username } = this.props.match.params;
```
Comment (collinewait) [inline src/containers/Profiles/UserProfileContainer.jsx]: `username` and `match` missing in prop validation. 
We can also use destructuring like:
` const { match: { params: {username} } = { params: {} } } = this.props;`

Use destructuring to get items on `this.props`. `this.props` is duplicated in this file. Make the appropriate changes.

File: src/containers/Profiles/UserProfileContainer.jsx [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+    };
+
+    componentDidMount = () => {
+        this.fetchProfile();
+    };
+
+    componentDidUpdate = prevProps => {
+        if(this.props.profile.username !== this.props.match.params.username){
+            this.fetchProfile();
+        };
```
Comment (collinewait) [inline src/containers/Profiles/UserProfileContainer.jsx]: Unnecessary semicolon

File: src/containers/Profiles/UserProfileContainer.jsx [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+
+    onDismissHandler = () => {
+        this.props.dismissMessage();
+    };
+
+    render() {
+        return (
+            <UserProfile {...this.props} onDismissHandler={this.onDismissHandler} />
+        );
+    };
```
Comment (collinewait) [inline src/containers/Profiles/UserProfileContainer.jsx]: Unnecessary semicolon

File: src/containers/Profiles/UserProfileContainer.jsx [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+};
```
Comment (collinewait) [inline src/containers/Profiles/UserProfileContainer.jsx]: Unnecessary semicolon

File: src/store/actions/async/profiles.js [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+
+export const fetchUserProfile = payload => {
+    return async dispatch => {
+        const { username, token } = payload;
+        dispatch(startedFetchUserProfile());
+        try {
+            const response = await axios.get(
+                `https://ah-backend-summer-staging.herokuapp.com/api/v1/profiles/${username}`, 
+                { headers: { Authorization: `Bearer ${token}` }}
+                );
```
Comment (collinewait) [inline src/store/actions/async/profiles.js]: `);` should be on the same line with `const`. Adjust the indentation.

File: src/store/actions/async/profiles.js [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+            const { image, first_name, last_name, bio } = data;
+            fd.append("image", image, image.name);
+            fd.append("first_name", first_name);
+            fd.append("last_name", last_name);
+            fd.append("bio", bio);
+            const response = await axios.patch(
+                `https://ah-backend-summer-staging.herokuapp.com/api/v1/profiles/${username}`, 
+                fd, 
+                { headers: {Authorization: `Bearer ${token}`}}
+                );
```
Comment (collinewait) [inline src/store/actions/async/profiles.js]: `);` should be on the same line with `const`. Adjust the indentation.

File: src/store/actions/async/profiles.js [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+            const responseData = response.data;
+            dispatch(updateUserProfile({responseData, message: "Edited profile."}));
+            localStorage.setItem("username", responseData.profile.username);
+            payload.history.push(`/${responseData.profile.username}`);
+        } catch (error) {;
```
Comment (collinewait) [inline src/store/actions/async/profiles.js]: Unnecessary semicolon

File: src/store/actions/async/profiles.js [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+            const message = "Something went wrong, Please check your internet connection."
```
Comment (collinewait) [inline src/store/actions/async/profiles.js]: Terminate this line

File: src/store/actions/sync/__tests__/profile.test.js [hunk 34]
```diff
@@ -0,0 +1,56 @@
...
+    it("failedFetchProfile creates action to set message when fetching profile fails", () => {
+        const action = { type: FAILED_FETCH_PROFILE, message: "" };
+        expect(failedFetchProfile("")).toEqual(action);
+    });
+
+    it("failedUpdateProfile creates action to set message when updating profile fails", () => {
+        const action = { type: FAILED_UPDATE_PROFILE, message: "" };
+        expect(failedUpdateProfile("")).toEqual(action);
+    });
+})
```
Comment (collinewait) [inline src/store/actions/sync/__tests__/profile.test.js]: EOF

File: src/store/reducers/profiles.js [hunk 35]
```diff
@@ -0,0 +1,60 @@
...
+            };
+        case FAILED_UPDATE_PROFILE:
+            return {
+                ...state,
+                loadingEditButton: false,
+                profileUpdateMessage: action.message
+            };
+        default:
+            return state;
+    };
```
Comment (collinewait) [inline src/store/reducers/profiles.js]: Unnecessary semicolon.

File: src/routes/index.jsx [hunk 36]
```diff
@@ -25,16 +28,20 @@ const Routes = () => (
...
                         <Route path="/articles/create" exact component={CreateArticle} />
                         <Route path="/articles/update/:slug" exact component={UpdateArticle} />
-                        <Route path="/articles/:slug" component={Article} />
+                        <Route path="/tags" component={LoginView} />
+                        <Route exact path="/:username" component={UserProfileContainer} />
+                        <Route exact path="/:username/edit" component={EditProfileContainer} />
+                        <Route exact path="/:username/followers" component={FollowersFollowing} />
+                        <Route exact path="/:username/following" component={FollowersFollowing} />
+                        <Route path="/email-verification/:token" exact component={EmailVerificationView} />
                     </Switch>
```
Comment (pbkabali) [inline src/routes/index.jsx:44]: @oma0256, this is nice work. the UI looks good. I think the switch for the URLs is not working as desired as I am now not able to get to the create articles page. Instead, it tries to look for article with slug: articles and returns NOT FOUND. Would you look into that please

**Review by missvicki — APPROVED** [review]
Thank you for working on the feedback I gave.

File: src/routes/index.jsx (hunk 36 above)
Comment (oma0256) [reply src/routes/index.jsx:44]: @pbkabali Thanks for bringing this to my attention let me look into this

**Review by missvicki — COMMENTED** [review]
Currently, new article is failing could you please look into that

**Comment by oma0256** [comment]
@missvicki @pbkabali managed to fix the issue of the new article

**Review by pbkabali — APPROVED** [review]

**Review by karuhanga — APPROVED** [review]
LGTM

**Pull request merged**